// Copyright 2022 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

syntax = "proto2";

option optimize_for = LITE_RUNTIME;

package apps.proto;

// This file is a mirror of the proto file maintained in the server code base at
// go/app-provisioning-service-proto. Changes should be made by updating the
// server code base and then copying the result to Chromium.
// In the future this file will be add only, however breaking changes can be
// be made until the interface is finalized (currently aimed for M110).

// TODO(b/243338003): Review and finalise API. >> THIS IS A DRAFT API ONLY <<

message AppProvisioningListAppsRequest {
  enum UserType {
    // Default for deserialization when an unexpected value is encountered.
    // Indicates to the client that the server has a new platform and needs
    // the proto file updated.
    USERTYPE_UNKNOWN = 0;

    // Normal consumer.
    USERTYPE_UNMANAGED = 1;

    // Enterprise, education, etc.
    USERTYPE_MANAGED = 2;

    // A child account that has restricted access to adult content.
    USERTYPE_CHILD = 3;

    // A guest account on the device.
    USERTYPE_GUEST = 4;
  }

  enum Channel {
    // Default for deserialization when an unexpected value is encountered.
    // Indicates to the client that the server has a new platform and needs
    // the proto file updated.
    CHANNEL_UNKNOWN = 0;

    // Channel used for test images, local dev, etc. where a channel is not
    // explicitly defined.
    // Note: Confusingly on the client-side ChromeOS calls this "unknown"
    CHANNEL_UNDEFINED = 1;

    // Canary channel.
    CHANNEL_CANARY = 2;

    // Developer channel.
    CHANNEL_DEV = 3;

    // Beta channel.
    CHANNEL_BETA = 4;

    // Stable channel.
    CHANNEL_STABLE = 5;
  }

  // The board identifier for the device sending the request.
  optional string board = 1;

  // The model identifier for the device sending the request.
  optional string model = 2;

  // The SKU identifier for the device sending the request.
  optional string sku_id = 3;

  // The language-country identifier for the user in language in
  // "language-COUNTRY" format, and must match one of the valid Google
  // recognised codes. See:
  // //i18n/identifiers/languagecodeconverter.h
  // Note: this should ideally be the user's preferred language, if absent
  // use the language specified by the UI. The server will attempt to honour
  // this explicitly, but if it cannot it wil fall-back onto the next best
  // available or en-US in the worst.case.
  optional string language = 4;

  // The ChromeOS version information.
  optional ChromeOSVersion chrome_os_version = 5;

  message ChromeOSVersion {
    // Ash Chrome version.
    optional string ash_chrome = 1;

    // Chrome OS platform version.
    optional string platform = 2;

    // The channel the device is using.
    optional Channel channel = 3;
  }

  // The type of user account making this request.
  optional UserType user_type = 6;
}

message AppProvisioningListAppsResponse {
  // A list of zero or more apps for APS to install.
  repeated App apps_to_install = 1;

  enum InstallReason {
    // Default for deserialization when an unexpected value is encountered.
    // Indicates to the client that the server has a new reason and needs
    // the proto file updated.
    INSTALL_REASON_UNKNOWN = 0;

    // A Default App.
    INSTALL_REASON_DEFAULT = 1;

    // An app installed for an OEM.
    INSTALL_REASON_OEM = 2;
  }

  message Icon {
    // Url to query to get the icon. This will always be from the host
    // meltingpot.googleusercontent.com.
    optional string url = 1;

    // Width of the icon in pixels. While App icons are typically square
    // note there is no guarantee the image provided will be.
    optional uint32 width_in_pixels = 2;

    // Mime type of the icon.
    optional string mime_type = 3;

    // Whether or not we have permission from the platform to mask the icon.
    optional bool is_masking_allowed = 4;
  }

  // For Android-only metadata.
  message AndroidExtras {}

  // For Web-only metadata.
  message WebExtras {
    // A URL to the web app's manifest in json format. This will always be from
    // the host meltingpot.googleusercontent.com.
    optional string manifest_url = 1;

    // The URL for the original source used to retrieve the contents of the
    // manifest above. This is needed to resolve the file on the client-side.
    optional string original_manifest_url = 2;
  }

  message App {
    // The primary key for the installer. This will always be of the format
    // "platform:primary_key". For now we support "android:package_name" and
    // "web:http://manifest/id".
    optional string package_id = 1;

    // The identifier for the App Group that Fondue used to derive this app
    // instance in the response.
    // Note: this may not be unique in the apps_to_install collection.
    optional string app_group_uuid = 2;

    // The App's UTF-8 encoded name in the requested language (or next best).
    optional string name = 3;

    // One or more Icons for this App for the requested language (or next best).
    repeated Icon icons = 4;

    // The reason why this app is in the list.
    optional InstallReason install_reason = 5;

    // Every platform has its own [Platform]Extras message to store platform
    // specific metadata.
    oneof extras {
      AndroidExtras android_extras = 6;
      WebExtras web_extras = 7;
    }
  }
}
